ReactNativeClient/lib/services/searchengine/SearchEngine.js

```diff
@@ -432,54 +448,45 @@ class SearchEngine {
 		return SearchEngine.SEARCH_TYPE_FTS;
 	}
 
-	async search(query, options = null) {
+	async search(searchString, options = null) {
 		options = Object.assign({}, {
 			searchType: SearchEngine.SEARCH_TYPE_AUTO,
 		}, options);
 
-		query = this.normalizeText_(query);
+		searchString = this.normalizeText_(searchString);
 
-		const searchType = this.determineSearchType_(query, options.searchType);
-		const parsedQuery = this.parseQuery(query);
+		const searchType = this.determineSearchType_(searchString, options.searchType);
 
 		if (searchType === SearchEngine.SEARCH_TYPE_BASIC) {
 			// Non-alphabetical languages aren't support by SQLite FTS (except with extensions which are not available in all platforms)
-			const rows = await this.basicSearch(query);
+			const rows = await this.basicSearch(searchString);
+			const parsedQuery = this.parseQuery(searchString);
 			this.processResults_(rows, parsedQuery, true);
 			return rows;
-		} else { // SEARCH_TYPE_FTS
+		} else {
+			// SEARCH_TYPE_FTS
 			// FTS will ignore all special characters, like "-" in the index. So if
 			// we search for "this-phrase" it won't find it because it will only
 			// see "this phrase" in the index. Because of this, we remove the dashes
 			// when searching.
 			// https://github.com/laurent22/joplin/issues/1075#issuecomment-459258856
-			query = query.replace(/-/g, ' ');
-
-			// Note that when the search engine index is somehow corrupted, it might contain
-			// references to notes that don't exist. Not clear how it can happen, but anyway
-			// handle it here by checking if `user_updated_time` IS NOT NULL. Was causing this
-			// issue: https://discourse.joplinapp.org/t/how-to-recover-corrupted-database/9367
```

Comment:
What do we do about this?

Adding `AND notes_fts.user_updated_time IS NOT NULL` to every FTS query will slow it down by about 10 times.

Reply:
In your queries, what will happen if the search index references a note that doesn't exist? Normally we should fix the underlying cause but since we don't know what it is, we should make sure the app doesn't break in the meantime.

For sure don't add this condition everywhere, it should just be in one place. If it slow down the queries a lot, then is there any other fast way to check? What if, before returning the rows, you quickly look through the rows and check for empty ones (and remove them)?